pkg/kubelet/cm/dra/plugin/noderesources.go

```diff
@@ -0,0 +1,449 @@
+
+	// resources is protected by the nodeResourcesController read/write lock.
+	// When receiving updates from the driver, the entire slice gets replaced,
+	// so it is okay to not do a deep copy of it. Only retrieving the slice
+	// must be protected by a read lock.
+	resources []*resourceapi.NodeResourceModel
+}
+
+// startNodeResourcesController constructs a new controller and starts it.
+//
+// If a kubeClient is provided, then it synchronizes NodeResourceSlices
+// with the resource information provided by plugins. Without it,
+// the controller is inactive.
+func startNodeResourcesController(ctx context.Context, kubeClient kubernetes.Interface, nodeName string) *nodeResourcesController {
+	if kubeClient == nil {
+		return nil
+	}
+
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithName(logger, "node resources controller")
+	ctx = klog.NewContext(ctx, logger)
+
+	c := &nodeResourcesController{
+		ctx:           ctx,
+		kubeClient:    kubeClient,
+		nodeName:      nodeName,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "node_resource_slices"),
+		activePlugins: make(map[string]*activePlugin),
+	}
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.run(ctx)
+	}()
+
+	return c
+}
+
+// waitForStop blocks until all background activity spawned by
+// the controller has stopped. The context passed to start must
+// be canceled for that to happen.
+//
+// Not needed at the moment, but if it was, this is what it would
+// look like...
+// func (c *nodeResourcesController) waitForStop() {
+// 	if c == nil {
+// 		return
+// 	}
+//
+// 	c.wg.Wait()
+// }
+
+// addPlugin is called whenever a plugin has been (re-)registered.
+func (c *nodeResourcesController) addPlugin(driverName string, pluginInstance *plugin) {
+	if c == nil {
+		return
+	}
+
+	klog.FromContext(c.ctx).V(2).Info("Adding plugin", "driverName", driverName)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if active := c.activePlugins[driverName]; active != nil {
+		active.cancel(errors.New("plugin has re-registered"))
+	}
+	active := &activePlugin{}
+	cancelCtx, cancel := context.WithCancelCause(c.ctx)
+	active.cancel = cancel
+	c.activePlugins[driverName] = active
+	c.queue.Add(driverName)
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		c.monitorPlugin(cancelCtx, active, driverName, pluginInstance)
+	}()
+}
+
+// removePlugin is called whenever a plugin has been unregistered.
+func (c *nodeResourcesController) removePlugin(driverName string) {
+	if c == nil {
+		return
+	}
+
+	klog.FromContext(c.ctx).V(2).Info("Removing plugin", "driverName", driverName)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if active, ok := c.activePlugins[driverName]; ok {
+		active.cancel(errors.New("plugin has unregistered"))
+		delete(c.activePlugins, driverName)
+		c.queue.Add(driverName)
+	}
+}
+
+func (c *nodeResourcesController) monitorPlugin(ctx context.Context, active *activePlugin, driverName string, pluginInstance *plugin) {
```

Comment:
Let#s add a comment about what this function does -- i.e. streams from ListAndWatch and tells the work queue to ...